### PR TITLE
fix: default collate_fn to pxt_torch_dataset_collate_fn in all GoldDoer classes

### LINKED ISSUE #202: when describing from table, multi labels is broken

### PR DESCRIPTION
**Pull request by AshleyAHuang**
Fixes #202. When collate_fn is left at its default (None), the
GoldDoers' DataLoaders fall through to PyTorch's own default_collate,
which transposes list-valued batch fields (e.g. a multi-label list)
across the batch dimension instead of keeping one list per sample.
Each sample silently loses all but one of its labels.

Fix (per the direction settled in #202): every GoldDoer's __init__ now
defaults collate_fn to pxt_torch_dataset_collate_fn when None —
GoldDescriptor, GoldVectorizer, GoldSelector, and GoldClusterizer —
and GoldSplitter.get_split_indices applies the same fallback for the
DataLoader it builds internally.

Companion fixes that this required:
- pxt_torch_dataset_collate_fn also stacks raw torch.Tensor, bool,
  float, and numpy-scalar values (previously only numpy.ndarray/int),
  since existing usage relies on tensor-valued fields with
  collate_fn=None.
- Sample-based schema inference now always goes through the collate,
  which adds a batch dimension — the clusterize/select inference sites
  missing a .squeeze(0) got one (select's dataset path already had it).

The simplification of the collate function itself (delegating to
torch's default_collate for non-sequence fields) is deliberately left
to #254.

Tests: regression test for the multi-label transposition (written
first, confirmed failing against the unfixed code), default-assignment
tests for all four GoldDoers, and a get_split_indices case with a None
selection value. Full suite: 423 passed; ruff + mypy clean.

### CODE REVIEW

**Comment by AshleyAHuang** [comment]
Good catch — pushed a fix. `bool` subclasses `int` in Python, so boolean fields were silently landing as `int64` tensors instead of `torch.bool`; float and numpy-scalar fields weren't handled at all and stayed as unbatched Python lists. Added explicit `bool`/`float`/`np.generic` branches (checked before `int` so booleans don't fall through) plus regression tests for each. Full suite still green (418 passed).

File: goldener/pxt_utils.py [hunk 1]
```diff
@@ -141,11 +144,16 @@ def stack_arrays_and_convert_to_torch(arrays: Sequence[np.ndarray]) -> torch.Ten
...
     for idx_sample, sample in enumerate(batch):
         for key, value in sample.items():
             value_list_dict[key].append(value)
             if idx_sample == 0:
-                if isinstance(value, np.ndarray):
+                if isinstance(value, torch.Tensor):
+                    conversion_dict[key] = stack_tensors
+                elif isinstance(value, np.ndarray):
                     conversion_dict[key] = stack_arrays_and_convert_to_torch
                 elif isinstance(value, int):
```
[the remark this answers is not in the document]
Comment (Yann-CV) [reply goldener/pxt_utils.py:155]: @AshleyAHuang If we want to match torch behavior we should directly use their collate utilities. I guess it would deserve its own pull request because I am pretty sure here it does not solve the issue with labels

File: goldener/pxt_utils.py [hunk 2]
```diff
@@ -129,26 +129,47 @@ def pxt_torch_dataset_collate_fn(batch: list[dict[str, Any]]) -> dict[str, Any]:
         batch: A list of samples, where each sample is a dictionary.
 
     Returns:
-        A single dictionary with collated values. The numpy arrays and integers
-        are stacked into torch tensors. All other types are kept as lists.
+        A single dictionary with collated values. The numpy arrays/scalars, booleans,
```
Comment (Yann-CV) [inline goldener/pxt_utils.py]: we can do another PR to modify this. here is my proposal (did not check typing):

```
def collate_mapping_batch_including_sequences(batch: list[dict[str, Any]])  -> dict[str, torch.Tensor | Sequence]:
    value_list_dict = defaultdict(list)
    conversion_dict: dict[str, Callable[[Sequence[Any]], torch.Tensor]] = {}

    for idx_sample, sample in enumerate(batch):
        for key, value in sample.items():
            value_list_dict[key].append(value)
            if idx_sample == 0:
                if not isinstance(value, Sequence):
                    conversion_dict[key] = default_collate

    return {
        key: conversion_dict[key](value) if key in conversion_dict else value
        for key, value in value_list_dict.items()
    }
```

Comment (Yann-CV) [reply goldener/pxt_utils.py]: by the way my bad I could have coded like that from the begining

File: goldener/describe.py [hunk 3]
```diff
@@ -490,7 +490,7 @@ def _sequential_describe(
             to_describe_dataset,
             batch_size=self.batch_size,
             num_workers=self.num_workers,
-            collate_fn=self.collate_fn,
+            collate_fn=self.collate_fn or pxt_torch_dataset_collate_fn,
```
Comment (Yann-CV) [inline goldener/describe.py]: should be done in the __init__ with logging info comment

Comment (Yann-CV) [reply goldener/describe.py]: by the way I guess we can generalize it to:
GoldClusterizer, GoldSelector, GoldVectorizer

I am wondering if we should not create our own Dataloader using pxt_torch_dataset_collate_fn as default collate_fn

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Thanks!

I am suggesting to modify the collate_fn in another PR (by the way we can simplify it by using directly the default pytorch one).

Also the same issue is also  present as well for other GoldDoers so either we create our own GoldDataLoader (and we use it in all our  GoldDoers) or we initialize the collate fn of all GoldDoer to our collate fn when the value is None (let's do it in the __init__).

**Comment by AshleyAHuang** [comment]
good call on simplifying — delegating to torch's default_collate for everything except list-valued fields is way cleaner than reimplementing type dispatch for tensor/int/float/bool/np.generic by hand. happy to rework it that way.

on the broader point (same bug in other GoldDoers) — that's a real architectural call, so want to check your preference before i just pick one: a shared GoldDataLoader wrapper used across all GoldDoers, or defaulting collate_fn to ours in each GoldDoer's __init__ when it's left as None? leaning toward the shared GoldDataLoader since it's one place to fix instead of N, but you'd know better if there's a reason the GoldDoers need to stay decoupled from a shared loader class. let me know which direction and i'll open the pr(s)

**Comment by Yann-CV** [comment]
> good call on simplifying — delegating to torch's default_collate for everything except list-valued fields is way cleaner than reimplementing type dispatch for tensor/int/float/bool/np.generic by hand. happy to rework it that way.

I created https://github.com/goldener-data/goldener/issues/254 for this! Feel free to handle it ;-)

**Comment by Yann-CV** [comment]
> leaning` toward the shared GoldDataLoader since it's one place to fix instead of N

@AshleyAHuang I guess you also have to move from DataLoader to GoldDataLoader in all the GoldDoers anyway.
I would go for the simpler which is, IMO replacing None by our own collate fn.

One pattern could be to create a GoldDoer superclass setting all the shared variable:

```
class GoldDoer:
  def __init__(
      self,
      allow_existing: bool = True,
      drop_table: bool = False,
      max_batches: int | None = None,
  ) -> None:
    ...


class GoldDoerWithTable(GoldDoer):
  def __init__(
      self,
      allow_existing: bool = True,
      drop_table: bool = False,
      max_batches: int | None = None,
      collate_fn: Callable | None = None,
      to_keep_schema: dict[str, type] | None = None,
      min_pxt_insert_size: int = 100,
      batch_size: int = 1,
      num_workers: int = 0,
  ) -> None:
      super().__init__(...)
      ... # do the collate update


then for instance
class GoldVectorizer(GoldDoerWithTable):
    def __init__(
        self,
        table_path: str,
        vectorizer: TensorVectorizer,
        collate_fn: Callable | None = None,
        data_key: str = "embeddings",
        target_key: str = "target",
        vectorized_key: str = "vectorized",
        label_key: str | None = None,
        target_to_label: dict[tuple[int, ...], str] | None = None,
        merge_multilabels: bool = False,
        exclude_labels: set[str] | None = None,
        exclude_full_zero_target: bool = False,
        to_keep_schema: dict[str, type] | None = None,
        min_pxt_insert_size: int = 100,
        batch_size: int = 1,
        num_workers: int = 0,
        allow_existing: bool = True,
        distribute: bool = False,
        drop_table: bool = False,
        max_batches: int | None = None,
    ) -> None:
        super().__init__(...)
        ...

```

Though we can as well keep it simple and do it in each class (which would not be a big deal I guess)

**Comment by AshleyAHuang** [comment]
reworked per the thread — collate_fn now defaults to pxt_torch_dataset_collate_fn in the `__init__` of all four GoldDoers instead of just patching GoldDescriptor's DataLoader. also gave `GoldSplitter.get_split_indices` the same fallback since it builds its own DataLoader.

one thing that tagged along: schema inference from a sample now always goes through the collate, which adds a batch dim — so the clusterize/select inference sites that were missing a `.squeeze(0)` got one (select's dataset path already had it). left the collate simplification itself out of here since that's #254.

full suite green locally (423 passed), ruff + mypy clean. branch is rebased on current main.

File: goldener/pxt_utils.py (hunk 2 above)
Comment (Yann-CV) [inline goldener/pxt_utils.py]: I would say here for now we ignore cubic feedback ==> PR is just about making sure we use our collat_fn

Comment (Yann-CV) [reply goldener/pxt_utils.py]: by the way potentially it removes temporary the need for the squeeze thing

File: tests/test_vectorize.py [hunk 4]
```diff
@@ -390,6 +391,23 @@ def setup_method(self):
     def teardown_method(self):
         pxt.drop_dir("unit_test", force=True)
 
+    def test_collate_fn_defaults_to_pxt_torch_dataset_collate_fn(self):
```
Comment (Yann-CV) [inline tests/test_vectorize.py:396]: here we should test the multilabel as well (like in tests/test_describe.py)

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Thanks @AshleyAHuang ! Good job! Quite close to the end ;-)

I would just suggest that we do not do the refactor on our collate_fn. After removing this changes, can you please check we still need the squeeze actions? (did not focused on it but I am not sure to see why the usage of our collate function would change anything).

**Comment by AshleyAHuang** [comment]
done! dropped the bool/float/np.generic handling — that was all the cubic-driven stuff. two things i couldn't drop, with receipts:

- **raw tensor stacking stays** (the one small collate addition left): the current collate keeps torch.Tensor fields as unstacked lists (it's only ever seen numpy from pixeltable datasets), so making it the default with zero changes = 67 test failures locally. one extra `isinstance(value, torch.Tensor)` branch fixes all of them.
- **the squeezes are still needed** — 11 failures without them. but you're right that it's not about *which* collate: it's the defaulting itself. before, `collate_fn=None` meant `get_dataset_sample_dict` / `get_sample_row_from_idx` skipped collation and schema inference saw a raw sample with no batch dim. now the collate always runs and adds one. that'll stay true under the #254 collate too, since default_collate also adds the batch dim.

also added the multilabel e2e test for GoldVectorizer like the describe one — double-checked it actually catches the bug (forcing default_collate: sample 0 only keeps class_1, sample 1 only class_2). suite green (423 passed), ruff + mypy clean.

**Review by Yann-CV — APPROVED** [review]
Great! Thanks a lot for the work!

**Comment by Yann-CV** [comment]
@AshleyAHuang I guess you also fixed https://github.com/goldener-data/goldener/issues/219 in the same time ;-)
